src/source/Dtls/Dtls.c

```diff
@@ -539,6 +546,7 @@ STATUS dtlsSessionPutApplicationData(PDtlsSession pDtlsSession, PBYTE pData, INT
     SIZE_T pending;
 
     MUTEX_LOCK(pDtlsSession->sslLock);
+    CHK(!ATOMIC_LOAD_BOOL(&pDtlsSession->shutdown), retStatus);
```

Comment:
Move this above the lock

Reply:
I cant, what could happen is the check passes and the thread block at the lock then shutdown happens.

Reply:
I see, no matter how you look at it, it's impossible to properly handle this for all cases. The situation is that if another thread is already shutting down the lock will be released after the shutdown. The main issue is to guard at the higher-layer logic against the object being freed before the check for the atomic.